lib/Dialect/HW/Transforms/HWConstProp.cpp

```diff
@@ -0,0 +1,334 @@
+   *
+   * `std::nullopt` is unknown, `IntegerAttr{}` is overdefined.
+   */
+  std::optional<IntegerAttr> map(Value value);
+
+private:
+  hw::InstanceGraph &graph;
+  DenseMap<Value, IntegerAttr> values;
+  DenseSet<Operation *> inQueue;
+  SmallVector<Operation *> overdefQueue;
+  SmallVector<Operation *> valueQueue;
+};
+} // namespace
+
+void ConstantPropagation::initialize(HWModuleOp module) {
+  if (module.isPublic()) {
+    // Mark public module inputs as overdefined.
+    for (auto arg : module.getBodyBlock()->getArguments())
+      markOverdefined(arg);
+  }
+
+  module.walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op)
+        .Case<ConstantOp>([&](auto cst) {
+          // Constants are omitted from the mapping, but their
+          // users are enqueued for propagation.
+          enqueue(cst, cst.getValueAttr());
+        })
+        .Case<HWInstanceLike>([&](auto inst) {
+          // Mark external/generated module outputs as overdefined.
+          bool hasUnknownTarget = llvm::any_of(
+              inst.getReferencedModuleNamesAttr(), [&](Attribute ref) {
+                Operation *referencedOp =
+                    graph.lookup(cast<StringAttr>(ref))->getModule();
+                auto module = dyn_cast_or_null<HWModuleOp>(referencedOp);
+                return !module;
+              });
+
+          if (hasUnknownTarget) {
+            for (auto result : inst->getResults())
+              markOverdefined(result);
+          }
+        })
+        .Case<hw::WireOp>([&](auto wire) {
+          // Mark wires as overdefined since they can be targeted by force.
+          markOverdefined(wire.getResult());
+        })
+        .Default([&](auto op) {
+          if (op->getNumResults() == 0)
+            return;
+          // Mark all non-comb ops and non-integer types as overdefined.
+          bool isFoldable = hw::isCombinational(op);
+          for (auto result : op->getResults()) {
+            Type ty = result.getType();
+            if (!hw::type_isa<IntegerType>(ty) || !isFoldable)
+              markOverdefined(result);
+          }
+        });
+  });
+}
+
+void ConstantPropagation::mark(Value value, IntegerAttr attr) {
+  auto it = values.try_emplace(value, attr);
+  if (!it.second) {
+    if (it.first->second == attr)
+      return;
+    attr = it.first->second = IntegerAttr{};
+  }
+  enqueue(value, attr);
+}
+
+void ConstantPropagation::enqueue(Value value, IntegerAttr attr) {
+  for (Operation *user : value.getUsers()) {
+    if (inQueue.insert(user).second) {
+      if (attr) {
+        valueQueue.push_back(user);
+      } else {
+        overdefQueue.push_back(user);
+      }
+    }
+  }
+}
+
+std::optional<IntegerAttr> ConstantPropagation::map(Value value) {
+  if (auto constant = value.getDefiningOp<hw::ConstantOp>())
+    return constant.getValueAttr();
+
+  auto it = values.find(value);
+  if (it == values.end())
+    return std::nullopt;
+
+  return it->second;
+}
```

Comment:
Calling this `map` seems a bit odd given that it's "getMaybeConstant". (That is a terrible name!) Something describing what this is doing would be better.